### ISSUE TITLE
Details re NNI models

### ISSUE THREAD

**Issue by pranabenator** [microsoft/nni#1351]
**What would you like to be added**: Details re NNI models, e.g. where are they located, how to test run them (with sample source code and test data).

**Why is this needed**: Training is not an end by itself ;-)

**Without this feature, how does current nni work**：Shows training accuracy to be very high but without testing, it is difficult to say.

**Components that may involve changes**: Unknown

**Brief description of your proposal if any**: None

**Comment by xuehui1991** [comment]
Hello @pablocastro. I don't think I totally understand your question. The "NNI models" means "the algorithm you choose from builtin-tuner" or "the trial models that you implemented"? 

How do you know "Shows training accuracy to be very high but without testing"? 

Thank you.

**Comment by pranabenator** [comment]
Suppose I follow your example of MNIST. It iterates 10 times and shows accuracy between 97 and 99%. However, how can I use the trained model? I couldn't find any mention in your docs. Secondly, I couldn't find any test code that would allow one to try out the model. Thirdly, how can I use NNI in bespoke projects?

**Comment by pranabenator** [comment]
@xuehui1991: you assigned issue raised by me i.e. @pranabenator to someone else i.e. @pablocastro. Is there any resolution in sight?

**Comment by pranabenator** [comment]
At least may I know if/when my concerns will be addressed? I am *unable to practically use* any part of NNI.

**Comment by xuehui1991** [comment]
Hello @pranabenator. 

In MNIST example, you just need to get the best-params (with high acc) in nni-board, and reset this best-params to the LeNet5 models and train from scratch, you could get the well-trained model.

But I guess you want to use nni to tune your DNN model right? And you want to know how to find the path of your trained model?

Actually, nni won't save the model for you. But you could add this code in trial by yourself.
For example, in tensorflow:
```python
# save models
saver = tf.train.Saver()
saver.save(sess, os.path.join(params['trial_id_x_save_path'], 'model.ckpt'))
# load models
tf.init_from_checkpoint(params['trial_id_x_restore_path'])
```

Also, you need to know the mapping relationship between param-config and  trial_id. So you may need to log this information.

Thank you.

**Comment by pranabenator** [comment]
Hi,
Could I use NNI with preprocessing pipelines? If so, how?
How to create a model with preprocessing integrated?
How did you preprocessing MNIST? Did you use the images or the 700+
features vector?
How could I get more information from you?
Do you support Microsoft partners?
Thanks,
-Pranab

**Comment by xuehui1991** [comment]
Hello @pranabenator. 
NNI is designed for hyper-parameter tuning or neural architecture search or any task that you try to use "search" to get an optimized solution.

So I not sure the preprocessing is a task that needs "search" to get a solution.
But if you want to integrate "preprocessing" into some trial code like MNIST, I think you could put it into trial code. But every trial will do the preprocessing logic and it will waste your time.

So I recommend you could maintain one preprocessed data and read them every time in each trial code.

You could use gitter to contact with us because it more like a chatroom and it may more easy to communicate.  We are happy to help all the community to use NNI.

Thank you.



Thank you for your question.

**Comment by pranabenator** [comment]
Responded via Gitter, thank you. -@pranabenator.

**Comment by pranabenator** [comment]
I have been corresponding with xuehui via email and Gitter, as suggested by them. The issue is not closed. Please help resolving: (1) What are the input to/output from NNI: where is the features matrix (if preprocessing is not integrated) and trained model? (2) If preprocessing pipeline can be integrated with the model like ML.Net?

I believe, without full disclosure of the aforesaid, one cannot use NNI for their bespoke project. I was told (via email) that they didn't integrate preprocessing since it would run whenever the model changes. However, this is not necessarily true: one may hold the feature matrix as a static datastructure or a singleton pattern so that it will be extracted/processed only when the variable is not initialized. Even if preprocessing is excluded, where are features being read from?

**Comment by QuanluZhang** [comment]
> **What would you like to be added**: Details re NNI models, e.g. where are they located, how to test run them (with sample source code and test data).
> 
> **Why is this needed**: Training is not an end by itself ;-)
> 
> **Without this feature, how does current nni work**：Shows training accuracy to be very high but without testing, it is difficult to say.
> 
> **Components that may involve changes**: Unknown
> 
> **Brief description of your proposal if any**: None

@pranabenator according to my understanding of the feature request above is that you want automatic model saving and serving. And, yes, "training is not an end by itself". 

For model saving, @xuehui1991 has provided a concrete solution, that is, you could save your model in your trial code (for example, save it under the path `NNI_OUTPUT_DIR`).

For model serving, the current version of NNI does not support this part, NNI helps you find the best hyperparameters or neural architectures. Then you can serve your model with the found hyperparameters or the neural architecture by yourself.

I agree this is a good feature, we will put it in our backlog, but cannot have a concrete plan for it for now.

**Comment by pranabenator** [comment]
The whole of NNI is not very open: clear documentation is something very important for induction. What is the namespace of NAS? What are the types of architectures that it can perform hyperparamter tuning on? The NNI is a component with not very clearly documented/specified API or SPI is what makes things difficult for the uninitiated. For better community acceptance, can Microsoft pay a bit of attention to these areas? I have designed quite a few architectures based on Microsoft Azure services that include AI. I am eager to lay my hands on NNI to understand what it is capable of. Any help/metadata will be appreciated.

**Comment by QuanluZhang** [comment]
@pranabenator thanks for your valuable suggestion, we will improve the document very soon.
